src/Context.php

```diff
@@ -276,6 +142,23 @@ public function offsetUnset(mixed $offset): void
         throw new \LogicException('Context is immutable.');
     }
 
+    private function with(mixed ...$args): self
```

Comment:
Did you consider putting all the parameters from the constructor here, with their type, but then nullable, with default null value?

It would still allow you to only use named parameters, but now you have type safety. 

Instead of an array you reference the parameter directly and fallback to the original value using ??.

Reply:
Yes but did not succeed to do that. There is not way in PHP to check is null is passed or if it's the default value

Reply:
Ah, it doesn't work when a property is nullable. Forgot about that.

Too bad!